parsers/IN_PB.py

```diff
@@ -5,7 +5,8 @@
 from collections import defaultdict
 
 
-GENERATION_URL = "http://www.pstcl.org:9091/scadadata/pbGenData2"
+GENERATION_URL = "https://sldcapi.pstcl.org/wsDataService.asmx/pbGenData2"
+DATE_URL = "https://sldcapi.pstcl.org/wsDataService.asmx/dynamicData"
```

Comment:
Thanks for trying to fix this. `https://sldcapi.pstcl.org/wsDataService.asmx/pbGenData2` fails when I try this.

`poetry run test_parser 'IN-PB' production`

Reply:
Can you be more specific about it failing? It's a classic - "works for me locally". Double checking with `curl` the url is accessible and data format correct.

```
$ curl -v https://sldcapi.pstcl.org/wsDataService.asmx/pbGenData2
*   Trying 103.118.160.105:443...
* TCP_NODELAY set
* Connected to sldcapi.pstcl.org (103.118.160.105) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES256-GCM-SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=*.pstcl.org
*  start date: Jun 23 00:00:00 2021 GMT
*  expire date: Jun 23 23:59:59 2022 GMT
*  subjectAltName: host "sldcapi.pstcl.org" matched cert's "*.pstcl.org"
*  issuer: C=GB; ST=Greater Manchester; L=Salford; O=Sectigo Limited; CN=Sectigo RSA Domain Validation Secure Server CA
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x55ba7893ae10)
> GET /wsDataService.asmx/pbGenData2 HTTP/2
> Host: sldcapi.pstcl.org
> user-agent: curl/7.68.0
> accept: */*
> 
* Connection state changed (MAX_CONCURRENT_STREAMS == 100)!
< HTTP/2 200 
< cache-control: private, max-age=0
< content-type: application/json; charset=utf-8
< server: Microsoft-IIS/10.0
< x-aspnet-version: 4.0.30319
< x-powered-by: ASP.NET
< access-control-allow-headers: accept, content-type
< access-control-allow-origin: *
< access-control-allow-methods: POST, GET, OPTIONS
< date: Wed, 29 Sep 2021 08:31:04 GMT
< content-length: 7216
< 
{"grossGeneration":{"dateTimeWrongFormat":null,"value":2715.0,"flag":null,"ddeItem"...
```

Reply:
So running `poetry run test_parser 'IN-PB' production` works on your machine?

Reply:
So what you're really trying to get at is that the datetime was not returned correctly. It would have been easier to lead with this...

```
$ python test_parser.py IN-PB production
2021-09-29 10:27:49,667 DEBUG    urllib3.connectionpool         Starting new HTTPS connection (1): sldcapi.pstcl.org:443
2021-09-29 10:27:50,547 DEBUG    urllib3.connectionpool         https://sldcapi.pstcl.org:443 "GET /wsDataService.asmx/pbGenData2 HTTP/1.1" 200 7219
2021-09-29 10:27:50,837 DEBUG    urllib3.connectionpool         https://sldcapi.pstcl.org:443 "GET /wsDataService.asmx/dynamicData HTTP/1.1" 200 158
...
  File "test_parser.py", line 86, in test_parser
    assert all([type(e['datetime']) is datetime.datetime for e in res_list]), \
AssertionError: Datetimes must be returned as native datetime.datetime objects
```

After

```
$ python test_parser.py IN-PB production
2021-09-29 10:30:49,196 DEBUG    urllib3.connectionpool         Starting new HTTPS connection (1): sldcapi.pstcl.org:443
2021-09-29 10:30:50,043 DEBUG    urllib3.connectionpool         https://sldcapi.pstcl.org:443 "GET /wsDataService.asmx/pbGenData2 HTTP/1.1" 200 7220
2021-09-29 10:30:50,339 DEBUG    urllib3.connectionpool         https://sldcapi.pstcl.org:443 "GET /wsDataService.asmx/dynamicData HTTP/1.1" 200 159
Parser result:
[{'datetime': datetime.datetime(2021, 9, 29, 14, 58, 54, tzinfo=tzfile('/usr/share/zoneinfo/Asia/Kolkata')),
  'production': {'biomass': 25.3936348, 'coal': 1912.7150261689999, 'hydro': 395.1682741, 'solar': 295.430603},
  'source': 'punjasldc.org',
  'storage': {},
  'zoneKey': 'IN-PB'}]
---------------------
took 1.15s
min returned datetime: 2021-09-29T09:28:54+00:00 UTC
max returned datetime: 2021-09-29T09:28:54+00:00 UTC  -- OK, <2h from now :) (now=2021-09-29T09:30:50.341992+00:00 UTC)
```

Reply:
When I tested this with `poetry run test_parser 'IN-PB' production`, then `https://sldcapi.pstcl.org/wsDataService.asmx/pbGenData2` failed, so it never failed with the `Datetimes must be returned as native datetime.datetime objects` error, hence why I didn't write it here.

Likely the website was down when I tested. :)

Reply:
I've tested from our infrastructure servers, and here the request goes through. So likely they are blocking some Danish IPs.